crates/holochain_persistence_api/src/txn/mod.rs

```diff
@@ -0,0 +1,166 @@
+/// Transactional trait extensions to the CAS and EAV persistence
+use crate::{cas::content::*, cas::storage::ContentAddressableStorage, eav::*, error::*};
+use std::{
+    collections::BTreeSet,
+    marker::PhantomData,
+    sync::{Arc, RwLock},
+};
+
+/// Defines a transactional writer, typically implemented over a cursor.
+pub trait Writer {
+    /// Commits the transaction. Returns a `PersistenceError` if the
+    /// transaction does not succeed.
+    fn commit(&mut self) -> PersistenceResult<()>;
+
+    /// Aborts the transaction explicitly. Otherwise, called upon `drop`.
+    fn abort(&mut self) -> PersistenceResult<()>;
+}
+
+/// Cursor interface over both CAS and EAV databases. Provides transactional support
+/// by providing a `Writer` across both of them.
+pub trait Cursor<A: Attribute>: Writer {
+    /// Adds the given EntityAttributeValue to the EntityAttributeValueStorage
+    /// append only storage.
+    fn add_eavi(
+        &mut self,
+        eav: &EntityAttributeValueIndex<A>,
+    ) -> PersistenceResult<Option<EntityAttributeValueIndex<A>>>;
+
+    /// Fetch the set of EntityAttributeValues that match constraints according to the latest hash version
+    /// - None = no constraint
+    /// - Some(Entity) = requires the given entity (e.g. all a/v pairs for the entity)
+    /// - Some(Attribute) = requires the given attribute (e.g. all links)
+    /// - Some(Value) = requires the given value (e.g. all entities referencing an Address)
+    fn fetch_eavi(
+        &mut self,
+        query: &EaviQuery<A>,
+    ) -> PersistenceResult<BTreeSet<EntityAttributeValueIndex<A>>>;
+
+    /// adds AddressableContent to the ContentAddressableStorage by its Address as Content
+    fn add(&mut self, content: &dyn AddressableContent) -> PersistenceResult<()>;
```

Comment:
Why `&dyn`? 

```suggestion
    fn add<C: AddressableContent>(&mut self, content: &C) -> PersistenceResult<()>;
```

Reply:
Original api had this, will probably just compose the existing content addressable storage and eavi traits instead.